hypothesis-python/tests/cover/test_functions.py

```diff
@@ -125,3 +129,61 @@ def test_functions_strategy_with_kwonly_args(f):
         f(1, 2)
     f(1, kwonly_arg=2)
     f(kwonly_arg=2, arg=1)
+
+
+def pure_func(arg1, arg2):
+    ...
+
+
+@given(
+    f=functions(like=pure_func, returns=integers(), pure=True),
+    arg1=integers(),
+    arg2=integers()
+)
+def test_functions_pure_with_same_args(f, arg1, arg2):
+    assert f(arg1, arg2) == f(arg1, arg2)
+
+
+@given(
+    f=functions(like=pure_func, returns=integers(), pure=True),
+    arg1=integers(),
+    arg2=integers()
+)
+def test_functions_pure_with_different_args(f, arg1, arg2):
+    r1 = f(arg1, arg2)
+    r2 = f(arg2, arg1)
+    assume(r1 != r2)
+    assert r1 != r2
+
+
+@given(
+    f1=functions(like=pure_func, returns=integers(), pure=True),
+    f2=functions(like=pure_func, returns=integers(), pure=True)
+)
+def test_functions_pure_two_functions_different_args_same_result(f1, f2):
+    r1 = f1(1, 2)
+    r2 = f2(0, 0)
+    assume(r1 == r2)
+    assert r1 == r2
+
+
+@given(
+    f1=functions(like=pure_func, returns=integers(), pure=True),
+    f2=functions(like=pure_func, returns=integers(), pure=True)
+)
+def test_functions_pure_two_functions_same_args_same_result(f1, f2):
+    r1 = f1(0, 0)
+    r2 = f2(0, 0)
+    assume(r1 == r2)
+    assert r1 == r2
+
+
+@given(
+    f1=functions(like=pure_func, returns=integers(), pure=True),
+    f2=functions(like=pure_func, returns=integers(), pure=True)
+)
+def test_functions_pure_two_functions_different_args_different_result(f1, f2):
```

Comment:
The property of interest here is in fact `test_functions_pure_two_functions_same_args_different_result` - which would fail if we had forgotten to include the generated function we're calling in the shared key.